### DISCUSSION TITLE
getStaticProps return undefined during build

### DISCUSSION THREAD

**Discussion by rake5h1** [vercel/next.js#15944] [Help]
getStaticProps returns undefined while fetching data from external API.

**Comment by filipesmedeiros** [comment]
Can you please provide more information so we can help? Maybe a code excerpt? Is it doing the fetching asynchronously?

**Comment by rake5h1** [comment]
I am using getStaticPaths and getStaticPaths to fetch posts from external API.While development it works fine if we navigate to that page but when I reload it shows an error.getStaticProps returns undefined.

export async function getStaticPaths() {
  
  const res = await axios.get(`${process.env.URL}/recent`);
  const data = await res.data;
  const paths = data.map((post) => {
    return {
      params: { id: post.Title },
    };
  });

  return { paths, fallback: true };
}
export async function getStaticProps({ params }) {
  const id = params.id;
  const res = await axios.get(`${process.env.URL}/postid?title=${id} `);
  const data = await res.data;
  return {
    props: { data: data },
  };
}

It is weird because for one post it works reloading but for other it shows error.Also during build it shows an error.

**Reply by KevenAndrade** [reply]
in your getStaticProps function, instead of `res.data` use `{ data }`.

**Comment by filipesmedeiros** [comment]
Ok you have some weird things that I will assume are just typos here and are not in the actual code (like the axios argument strings not being inside quotes).

Also I'm not too familiar with axios but I think `res.data` is not a Promise.

Other than that, this seems fine, to be honest. What are the file names for the relevant paths on your project?

**Comment by rake5h1** [comment]
routes are /Post/[id]

**Comment by filipesmedeiros** [comment]
In that case the only thing I can ask/recommend is you log the axios responses and data to make sure it's coming correctly on both functions. Maybe the server is not returning what you expect? Sorry I couldn't help further

**Comment by rake5h1** [comment]
Server is returning data.Routes also work fine when i navigate through home page but if we reload the route then getSaticProps is returning undefined.getServerSideProps works fine.

**Reply by filipesmedeiros** [reply]
getStaticProps shouldn't fail at run time. It's a compile time function only. It fails when you're building it. I think in development it gets run everytime you GET the route. Can you try building to see if it works?

**Reply by rake5h1** [reply]
During build also it fails.getStaticProps returns undefined

**Reply by filipesmedeiros** [reply]
I see no other alternative than it being your API code. Are you absolutely sure the API is returning the correct stuff? Can you `console.log(res.data)` please? Also, is the source code open so I can check?

**Reply by rake5h1** [reply]
Yes. Check the c7ech-next repo master branch.
/Post/[id] route

**Reply by filipesmedeiros** [reply]
Everything looks good, I can't help anymore, sorry. Just `console.log` the data returned in both static functions to see if both are returning what they should, I guess.

**Reply by rake5h1** [reply]
data is returned if we navigate from the home page but if we reload the route then getStaticProps returns undefined.
console.log shows undefined.Redirects to 404 page

**Reply by filipesmedeiros** [reply]
What if you come from another page (like Google or whatever) and enter the post path directly?

**Reply by rake5h1** [reply]
It shows error even if we enter the path directly. The API is working fine I  have cross-checked with the postman and the route also works if we navigate to. During build also undefined is returned.
Can u suggest any other alternative for data fetching during build time?

**Reply by rake5h1** [reply]
Hey! The build is now working. And I figured during development if in getStaticPaths fallback is false then 404 is returned and if fallback is true getStaticProps first returns undefined and refreshes in the background and returns the data. It works.

Thanks For Your Help.

**Reply by markitics** [reply]
"during development... if getStaticPaths fallback true, getStaticProps first returns undefined, then refreshes in the background and returns the data" – thank you for updating the thread wit this comment! 
I've added `if (!data){return <div>Loading…</div>}` in my page view, and it works now!

**Reply by arturs706** [reply]
> "during development... if getStaticPaths fallback true, getStaticProps first returns undefined, then refreshes in the background and returns the data" – thank you for updating the thread wit this comment!
> I've added `if (!data){return <div>Loading…</div>}` in my page view, and it works now!

Hi
Where shall I put this?
Have the same issue.

**Comment by rake5h1** [comment]


**Reply by filipesmedeiros** [reply]
Not too familiar with caching, but is it even possible to cache an API route or dyamic page in Next?

Maybe try this? `Cache-Control: public,max-age='duration',immutable`?

**Comment by mhesham32** [comment]
you can use this when fallback is set to true 
```js
 const router = useRouter();

  if (router.isFallback) {
    return <h1>Loading...</h1>;
  }
```
put it at the very top of your component to avoid getting of undefined errors on the build time. before using the prop provided by getStaticProps.

**Reply by nishant-iyengar** [reply]
thank you, this helped me!

**Reply by pedegago** [reply]
This helped me too!
Also, I was searching for an explanation of this and found it: https://nextjs.org/docs/basic-features/data-fetching#fallback-pages

**Comment by masonbrothers** [comment]
I'm just putting this here if anyone is having a similar issue in 2022. 

I tried @mhesham32's solution, but it didn't work for my use case (as I don't have fallback pages). I would recommend trying their solution though to anyone before using mine as their's is better if you have fallback pages.

I think there are multiple renders, and `undefined` is passed to page props on the first render. On the second, the the actual value is passed.

You might be able to do this if it suits your use case:
```javascript
function ProductPage({ product}) {
  if (typeof product === "undefined") return "Loading..."
...
```
